packages/runtime/container-runtime/src/blobManager.ts

```diff
@@ -61,40 +62,116 @@ export interface IBlobManagerLoadInfo {
 
     constructor(
         private readonly routeContext: IFluidHandleContext,
         snapshot: IBlobManagerLoadInfo,
         private readonly getStorage: () => IDocumentStorageService,
-        private readonly attachBlobCallback: (blobId: string) => void,
+        private readonly sendBlobAttachOp: (blobId?: string, localId?: string) => void,
         // To be called when a blob node is requested. blobPath is the path of the blob's node in GC's graph. It's
         // of the format `/<BlobManager.basePath>/<blobId>`.
         private readonly gcNodeUpdated: (blobPath: string) => void,
-        private readonly runtime: IContainerRuntime,
-        private readonly logger: ITelemetryLogger,
+        private readonly runtime: IBlobManagerRuntime,
     ) {
-        this.runtime.once("dispose", () => {
-            for (const promise of this.pendingBlobIds.values()) {
-                promise.reject(new Error("runtime disposed while blobAttach op in flight"));
+        this.logger = ChildLogger.create(this.runtime.logger, "BlobManager");
+        this.runtime.on("disconnected", () => this.onDisconnected());
+        this.redirectTable = this.load(snapshot);
+    }
+
+    /**
+     * Upload blobs added while offline. This must be completed before connecting and resubmitting ops.
+     */
+    public async onConnected() {
+        await PerformanceEvent.timedExecAsync(
+            this.logger,
+            {
+                eventName: "BlobUploadOnConnected",
+                count: Array.from(this.pendingBlobs.values())
+                    .filter((entry) => entry.status === PendingBlobStatus.OfflinePendingUpload).length,
+            },
+            async () => {
+                // if blobs are added while uploading, upload them too
+                while (this.hasPendingUploads) {
+                    const uploadPs: Promise<void>[] = [];
+                    for (const entry of this.pendingBlobs.values()) {
+                        assert(entry.status !== PendingBlobStatus.OnlinePendingUpload,
+                            "Must submit BlobAttach op before uploading offline blob");
+                        if (entry.status === PendingBlobStatus.OfflinePendingUpload) {
+                            uploadPs.push(this.getStorage().createBlob(entry.blob).then((response) => {
+                                entry.status = PendingBlobStatus.OfflinePendingOp;
+                                entry.storageId = response.id;
+                            }));
+                        }
+                    }
+                    await Promise.all(uploadPs);
+                }
+            },
+            { start: true, end: true },
+        );
+    }
+
+    private onDisconnected() {
+        // switch all blobs that haven't been uploaded yet to offline (local ID) flow
+        for (const [id, entry] of this.pendingBlobs) {
+            if (entry.status === PendingBlobStatus.OnlinePendingUpload ||
```

Comment:
That makes sense. So you're saying we should only transition to local/offline flow if the upload request fails and runtime is disconnected, otherwise just keep retrying? 

It seems like tracking the response would add complexity to the design, do you think it's worth trying to do this in this PR? Since the driver limits to 1 blob upload at a time, the effect would just be restarting the blob upload currently in progress.

Reply:
I've made some changes based on your feedback, and I also updated the PR description with more detail based on these changes. Please take a look!